Source/Core/DolphinQt2/Resources.cpp

```diff
@@ -2,46 +2,69 @@
 // Licensed under GPLv2+
 // Refer to the license.txt file included.
 
+#include <algorithm>
+
+#include <QGuiApplication>
+#include <QScreen>
 #include <QStringList>
 
 #include "Common/CommonPaths.h"
 #include "Common/FileUtil.h"
 #include "DolphinQt2/Resources.h"
+#include "DolphinQt2/Settings.h"
 
 QList<QPixmap> Resources::m_platforms;
 QList<QPixmap> Resources::m_countries;
 QList<QPixmap> Resources::m_ratings;
 QList<QPixmap> Resources::m_misc;
 
+QString Resources::GetPreferredIconScale(int max)
+{
+  const auto dpr = QGuiApplication::primaryScreen()->devicePixelRatio();
+  const auto scale = std::min(dpr <= 1 ? 1 : (dpr <= 2) ? 2 : 4, max);
+
+  if (scale > 1)
+    return QStringLiteral("@%1x").arg(scale);
+  else
+    return QStringLiteral("");
+}
+
+QPixmap Resources::GetScaledIcon(const QString& name, int max, const QString& dir)
+{
+  return QPixmap(QStringLiteral("%1%2%3.png")
+                     .arg(dir.isEmpty() ? Settings().GetResourcesDir() : dir, name,
+                          GetPreferredIconScale(max)));
+}
+
+QPixmap Resources::GetScaledThemeIcon(const QString& name, int max)
+{
+  return Resources::GetScaledIcon(name, max, Settings().GetThemeDir());
+}
+
 void Resources::Init()
 {
   QString sys_dir = QString::fromStdString(File::GetSysDirectory() + RESOURCES_DIR + DIR_SEP);
 
-  QStringList platforms{QStringLiteral("Platform_Gamecube.png"), QStringLiteral("Platform_Wii.png"),
-                        QStringLiteral("Platform_Wad.png"), QStringLiteral("Platform_File.png")};
+  QStringList platforms{QStringLiteral("Gamecube"), QStringLiteral("Wii"), QStringLiteral("Wad"),
+                        QStringLiteral("File")};
   for (QString platform : platforms)
-    m_platforms.append(QPixmap(platform.prepend(sys_dir)));
+    m_platforms.append(GetScaledIcon(platform.prepend(QStringLiteral("Platform_"))));
```

Comment:
I don't think that splitting e.g. "Platform_Gamecube" into "Platform_" and "Gamecube" is a good idea. It hurts greppability of the codebase, and there's no common logic here other than the filenames looking similar.